### DISCUSSION TITLE
Native Image Committer and Community Meeting 2024-11-28

### DISCUSSION THREAD

**Discussion by wirthi** [oracle/graal#10068] [Show and tell]
List of all past and upcoming meetings: #3933

Next meeting is planned for December 19, 2024. After that, there will be a meeting on the last Thursday of each month.

## New and Noteworthy

* Native Image Layers:
  #9994 
  #9962
  #9959 (90%-95% smaller `.nil` files, much faster builds)
  #10007
  #10146
  #10144
  #10119
  #10102
  #10100
  #10090
  #10067
  #10061
  #9995
  #9965
  #9953
  #9905
  
* GuestGraal
  #10082
  #10075
  #10049
  #9977

* Streamline reachability metadata
  #10178
  #10104
  #10180 
  
* Usability:
  #10053
  #10143
  #10030
  #9903
  https://github.com/oracle/graalvm-reachability-metadata/pull/552
  https://github.com/graalvm/native-build-tools/pull/624

* Monitoring / tools:
  #10129
  #10059
  #9963 (contributed by @roberttoyonaga from Rad Hat)
  
* Debugging
  #10112
  #9948
  #9900

* Performance and Image Size
  #9821 Whole-Program Sparse Conditional Constant Propagation (Skipflow) enabled by default in JDK25 (after our branch-off in December). Presentation in the next meeting Dec 19th!
  #9938

* Other
  #9883
  #9950
  #10076
  #10166
  #10156
  #10072
  #10169
  #9947
  

## Questions
@wirthi the Graal project favors rebasing over merging (to update to newer version). When adopting community PRs that causes troubles (changed commit hashes, sometimes mirroring breaks). How much of a problem is that for the community; previously,we tried not to touch commits (not to change commit hashes) because that was more comfortable to the community.

### Summary
An implementation of the JCMD feature is being integrated. This will allow the JCMD tool to attach to a Native Image executable and issue diagnostic commands similar to what's possible in OpenJDK.

### Motivation
Native Image has a growing number of features such as NMT, heap dumps, thread dumps, and compilation dumps that rely on handling signals to perform operations requested at runtime. This is not ideal because there is a limited number of signals that are available to be used. It is also not ideal because arguments cannot be easily used with signals. Supporting JCMD would be a better alternative to signals because we can support an unlimited number of commands and also accept arguments.

JFR in OpenJDK heavily uses JCMD to accept user requests at runtime. Currently, there is no way to stop/start/dump recordings spontaneously at runtime (apart from using the remote JMX feature - which has significant overhead with respect to image size). This is inconvenient if a user wants to diagnose a problem without shutting down or restarting their application. For example, if a user wants to access JFR recordings from their application, they need to terminate their application so that a JFR dump is created. With JCMD they are able to dump recordings whenever they desire without needing to restart their entire application.

### General Implementation Notes
The JCMD feature relies on two components: the Attach API feature and the DCMD feature. These willa ll be integrated together. Implementation of the Attach API is what allows the JCMD tool to connect to the JVM, issue commands, and receive response data. The DCMD infrastructure is what parses and executes the individual diagnostic commands. 

There is a dedicated notification thread which handles Attach API connections and delegates command tasks to the DCMD feature. The Attach API uses unix domain sockets for communication. The Attach API is initialized with a handshake that involves the out of process client sending SIGQUIT and creating a file to indicate that initialization should begin. 

### Limitations
Not all JCMD diagnostic commands available in OpenJDK will initially be available in Native Image. Please see the documentation PR for an exact list of commands available: https://github.com/oracle/graal/pull/9987

Currently, this feature will not be available on Windows.  This is not really a regression since heap dump, thread dump, compilation dump, nmt dump, and JFR, all don't work on Windows anyway (since they require unix signals to operate).

The Windows implementation will be largely independent of code in this posix implementation. So there will be no need to go back and rework anything here. 

### Links
Internal PR https://github.com/oracle/graal/pull/9963
Original PR https://github.com/oracle/graal/pull/9232
Documentation PR https://github.com/oracle/graal/pull/9987

**Comment by zakkak** [comment]
Regarding:

> the Graal project favors rebasing over merging (to update to newer version). When adopting community PRs that causes troubles (changed commit hashes, sometimes mirroring breaks). How much of a problem is that for the community; previously,we tried not to touch commits (not to change commit hashes) because that was more comfortable to the community.

As mentioned in the call, my only concern is whether GitHub PRs will be able to properly track the rebased commits. We already see some PRs appearing as closed despite actually being merged, e.g.:

* https://github.com/oracle/graal/pull/9994
* https://github.com/oracle/graal/pull/9959
* https://github.com/oracle/graal/pull/10007

Not sure if this is related to rebasing/merging or to some issue with the graalvm-bot though. It's most probably the latter as if I go to https://github.com/oracle/graal/pull/10007/commits/8637bb7a6b623556402ad4b5cab1313022d6967a github shows that the commit is actually merged on `master`. One guess of mine is that when you merge multiple PRs in a single internal merge and then push the changes upstream only one of the upstream PRs ends up getting marked as merged and the others are marked as closed.

Another question is, if you end up rebasing upstream PRs would you also force push to the contributor's branch? To do so contributors need to check the "Allow edits and access to secrets by maintainers" checkbox when opening a PR. Another thing that needs to be taken in consideration is what happens if the contributor force pushes after the rebase? (I guess this is already an issue, since sometimes contributors force push or add new commits after their work has been integrated internally).

cc @vjovanov @wirthi

**Reply by christianhaeubl** [reply]
> We already see some PRs appearing as closed despite actually being merged

Thanks, I forwarded that internally.

> Another question is, if you end up rebasing upstream PRs would you also force push to the contributor's branch?

I think it makes sense to keep an unmodified version of the original contribution around. So, no, we would not force push to the contributor's branch. Unfortunately, this makes it even more likely that GitHub loses track. For more complex changes, the result would then probably look like the following:
- the final PR that was merged would have status "Merged"
- the original contributor PR would have status "Closed"
- both PRs would be linked (pretty much like we already do it now, see https://github.com/oracle/graal/pull/9232 for an example)

Would something like that be good enough?

> Another thing that needs to be taken in consideration is what happens if the contributor force pushes after the rebase? (I guess this is already an issue, since sometimes contributors force push or add new commits after their work has been integrated internally).

This is already a problem. When I start integrating a PR, I typically post a comment to indicate that no further changes should be pushed to this PR/branch. I think we should establish that as best practice.

**Reply by zakkak** [reply]
> * the final PR that was merged would have status "Merged"
> * the original contributor PR would have status "Closed"
> * both PRs would be linked (pretty much like we already do it now, see https://github.com/oracle/graal/pull/9232 for an example)

> Would something like that be good enough?

Does this mean that for rebased changes a new PR will be mirrored on GitHub and a comment will be added to the both PRs indicating the relationship between the two? If so it should be good enough.

> This is already a problem. When I start integrating a PR, I typically post a comment to indicate that no further changes should be pushed to this PR/branch. I think we should establish that as best practice.

:+1:

**Reply by christianhaeubl** [reply]
> Does this mean that for rebased changes a new PR will be mirrored on GitHub and a comment will be added to the both PRs indicating the relationship between the two?

Yes, a new PR will be mirrored for the rebased changes.
Thanks, then let's try this approach for one of the next contributions.

**Reply by zakkak** [reply]
Another instance were @olpaw had to manually close the issue?

#10161 

I also forgot to mention on my original comment that when GitHub fails to properly mark the PR as closed, the commit also doesn't link back to the PR as shown below:

![image](https://github.com/user-attachments/assets/5e6a8e57-8e44-43c3-8367-5296a0d89abd)

**Reply by christianhaeubl** [reply]
@ansalond Is this scenario covered by the already existing issue?

**Comment by ansalond** [comment]
> Another instance were @olpaw had to manually close the issue?
> https://github.com/oracle/graal/pull/10161

I only had a quick look, but from what I can see I believe that [these commits](https://github.com/oracle/graal/pull/10161/commits) were squashed manually into https://github.com/oracle/graal/commit/7d964fab3b3b922a2d2141f98823cb6f4b488573 before merging the upstream PR. I can see how that might confuse GitHub, but we do not have a ticket to prevent that.

> I also forgot to mention on my original comment that when GitHub fails to properly mark the PR as closed, the commit also doesn't link back to the PR as shown below

Indeed, that link appears only when the corresponding GitHub PR is merged rather than closed. In your example:
- closed: https://github.com/oracle/graal/pull/8815
- merged: https://github.com/oracle/graal/pull/8913

We have an internal ticket to change our mirroring so that PRs appear as merged, but I cannot provide an ETA.
